design/036-package_id.md

```diff
@@ -0,0 +1,107 @@
+# Proposal: Conan 2.0 requirements effect in PackageID
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [036](https://github.com/conan-io/tribe/pull/36)  |
+| **Submitted**     | 2022-06-17                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+This proposal is about the effect that the dependencies have over the current package ``package_id``, but not about other factors, such as the package settings, options, configuration or other type of binary compatibility resulting from other factors other than dependencies.
+
+Conan 2.0 proposes a new default ``package_id`` that will be dynamic, based on two main cases:
+
+- When an application is linking a static library, or a shared library is linking a static library, or any library or application is linking a header-only library as an implementation detail, the resulting artifacts will be fully embedding/inlining the dependency artifacts, and they require in the general case to be re-built from sources when something in the dependency changes. This will be the "embed" case, and by default will use a "full_mode" policy, which means that the full dependency reference ``pkg/version@user/channnel#recipe_revision:package_id`` will be used and factored in the consumer ``package_id``. Recall that ``package_revision`` can no longer be part of the ``package_id`` as removed in proposal https://github.com/conan-io/tribe/pull/30
+- When a static library is depending on a static library, or a header-only library is depending on another library, the resulting artifacts in general do not embed/inline the dependency artifacts, and they don’t need to be rebuilt in many cases even if the dependency changes. For example, doing a patch in an implementation file .cpp of a static library does not require another static library that depends on it to be recompiled. This will be the "non embed" case and by default will use a "minor_mode" policy, which means requiring a rebuild if the major or minor of the dependency changes, but not requiring it if the patch of the dependency changes.
+
+Conan will be using the ``package_type`` information to define which of the above scenarios apply.
+
+Conan 2.0 will allow ``tool_requires`` (or any other ``requires`` with trait ``build=True``) to affect the ``package_id`` of their consumers. By default it will be effect will be ``None``, but users will be able to define it, globally via ``global.conf`` configuration, or per recipe in the ``requires(...., build=True, package_id_mode=xxxx)`` trait (being ``xxx`` any valid mode like ``minor_mode``, ``patch_mode``, ``full_mode``, etc)
```

Comment:
I am still confused , does it mean this https://github.com/conan-io/conan/issues/7876 will be solved with moving the NDK dependency from a build_requires to a tools_requires section ?

and I add that for example in the profile, and this ensures all packages I use come from the same NDK build? 

and does the global setting win over the recipe setting or vice versa?

Reply:
``tool_requires`` is mostly an alias for ``build_requires`` to better illustrate that it is something to be used for "tools", and not for regular libraries (typical abuse in Conan 1.X)

Yes, https://github.com/conan-io/conan/issues/7876 will be solved by this, by different means:
- Specifying directly in the ``self.tool_requires("ndk/version", package_id_mode="minor_mode")``
- Specifying the general policy for tool_requires in ``global_conf``: ``core.package_id:default_build_mode=minor_mode``. This approach affects all the packages that are built with a ``tool_requires``, even if it comes defined in a profile.

At the moment the ``global.conf`` settings is explicitly called a "default", it will have less priority than the recipe ones. It seems that when a recipe decides to explicit the modes of its dependencies, it is very aware of what is doing, and that should be respected in general. If the case for a global configuration that also overrides the recipe defined ones, that might be doable.

Reply:
Thank you for your patience in explaining me the details!